Fix KeyPath with default arg

zoecarver opened the pull request:
<!-- What's in this pull request? -->
This change fixes key paths that use default arguments in their subscript. For example: 
```swift
struct Foo {
  subscript(x: Int = 0) -> Int { x }
}

_ = \Foo.[]
```

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-11665](https://bugs.swift.org/browse/SR-11665).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

My only lingering question is if anyone has objections to removing the call to [`relabelParams`](https://github.com/apple/swift/compare/master...zoecarver:key-path-default-arg?expand=1#diff-69306f7b0822cf52e8c0639307ad7407R962). It seems like it does nothing.

zoecarver commented:
cc @xedin @jckarter

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -950,8 +950,17 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
   argsWithLabels.append(args.begin(), args.end());
+
+  // TODO: is this necessary?
   AnyFunctionType::relabelParams(argsWithLabels, argLabels);
```
@slavapestov Do you remember why do we need to do this? Shouldn't labels in `args` always match the ones retrieved from expression by `getCalleeDeclAndArgs`?

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -950,8 +950,17 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
 
+  // TODO: if the above is necessary, argsWithLabels should be renamed or should be initialized using the following:
+  //  argsWithLabels.reserve(args.size());
+  //  std::copy_if(args.begin(), args.end(), argsWithLabels.begin(),
```
We can't actually do conditional copy here because arguments without labels still have to be included in `argsWithLabels` list.

zoecarver commented on lib/AST/Expr.cpp:
```diff
@@ -2176,7 +2176,7 @@ void KeyPathExpr::Component::setSubscriptIndexHashableConformances(
   switch (getKind()) {
   case Kind::Subscript:
-    assert(hashables.size() == SubscriptSize);
+    SubscriptSize = hashables.size();
```
Yes, you're right. I meant for this to update `getSubscriptIndexHashableConformances` not `getSubscriptLabels`.

zoecarver commented on lib/AST/ASTContext.cpp:
```diff
@@ -2981,7 +2981,7 @@ bool AnyFunctionType::equalParams(CanParamArrayRef a, CanParamArrayRef b) {
 void AnyFunctionType::relabelParams(MutableArrayRef<Param> params,
                                     ArrayRef<Identifier> labels) {
-  assert(params.size() == labels.size());
+  assert(params.size() <= labels.size());
```
I removed the call to `relabelParams` so, this assertion can safely remain. If the call _was_ important, we can debug this further.

theblixguy commented:
Can you please squash your commits and format it according to the guidelines [here](https://swift.org/contributing/#commit-messages)?

xedin commented on include/swift/AST/Expr.h:
```diff
@@ -4916,7 +4916,10 @@ class KeyPathExpr : public Expr {
     union {
-      unsigned SubscriptSize;
+      struct {
+        unsigned subscripConformancestSize;
```
Typo - `subscriptConformancesSize`

xedin replied:
I'd suggest we rename this to `numLabels`, and `numConformances` which is more concise and `SubscriptSize` to `SubscriptInfo`

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -950,7 +950,6 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
   // Apply labels to arguments.
   SmallVector<AnyFunctionType::Param, 8> argsWithLabels;
   argsWithLabels.append(args.begin(), args.end());
-  AnyFunctionType::relabelParams(argsWithLabels, argLabels);
```
I'd suggest we leave this be for now since this is not related to the problem.

zoecarver replied:
Will do. What about the assertion?

xedin replied:
I think that assertions makes more sense in SILGen maybe to make sure that number of indexHashes retrieved matches the number of parameters...

zoecarver replied:
I mean, what should we do about [the assertion](https://github.com/apple/swift/blob/5f3b1da22a3d0b2f2d90540cbbd8336db7ca931c/lib/AST/ASTContext.cpp#L2982) that is triggered in `relabelParams`?

hamishknight replied:
An alternative solution would be to update the subscript component in CSApply with a new list of argument labels to include the default arguments. We already have to [iterate over the parameters to dig out the conformances here](https://github.com/apple/swift/blob/4980bc71ee96649c3ac835c123fb10a89b7f0689/lib/Sema/CSApply.cpp#L4646-L4659), so we could also use that loop to build up a new vector of argument labels.

Eventually, we should probably make `coerceCallArguments` return the individually coerced arguments along with their labels instead of bundling them into a paren or tuple expr, at which point we could use the labels returned from that.

xedin replied:
I like what @hamishknight is proposing, since `coerceCallArguments` could produce defaults we can't actually rely on the label array from subscript component.

zoecarver replied:
Sounds good, I'll work on that.

zoecarver replied:
Turns out I made a mistake. This isn't actually an issue. I misdiagnosed something earlier. I won't change anything here but, maybe I will make another patch to update `coerceCallArguments` as you suggested, @hamishknight.

xedin reviewed (commented):
Couple of minor comments and this would be ready to land.

zoecarver committed:
Fix support for subscript default arguments in KeyPath

zoecarver commented:
@theblixguy I squashed my commits. Usually, I do that before merging (via squash and merge), but would you like my commits to be squashed in the PR too? Also, should I try to provide a body in addition to title?

xedin commented:
@zoecarver Don't worry about it, I'll squash before merge.

xedin commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3549,7 +3549,30 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
+
+      if (auto *tupleExpr = dyn_cast<TupleExpr>(component.getIndexExpr())) {
+        size_t count = 0;
+        for (auto *element : tupleExpr->getElements()) {
```
Actually here is one more refactoring suggestion:

```c++
for (unsigned i = 0, n = tupleExpr->getNumElements(); i != n; ++i) {
  const auto &element = tupleExpr->getElement(i);
  if (auto *defaultArg = dyn_cast<DefaultArgumentExpr>(element)) {
    // do the thing
  }
}
```

theblixguy replied:
Should we be modifying the AST in SILGen like this? Maybe we could do this without modifying the AST here? or perhaps do this in CSApply perhaps.

xedin replied:
I'm not expert in SILGen and was is going on there so @jckarter should take a look at this. Looking at `CSApply` `coerceCallArguments` should build a new index expression for subscript component which should include `DefaultArgumentExpr`.

zoecarver replied:
Yes. It looks like the following for the above example:
```
(paren_expr implicit type='(Int)' location=run.swift:19:10 range=[run.swift:19:10 - line:19:11]
  (default_argument_expr implicit type='Int' location=run.swift:19:10 range=[run.swift:19:10 - line:19:10] default_args_owner=run.(file).Foo.subscript(_:)@run.swift:16:3 param=0))
```

zoecarver replied:
I'll try to move the above default argument parsing into `coerceCallArguments`.

zoecarver replied:
When I moved it to `coerceCallArguments` I got errors in `visitIdentityExpr` (because `getType` has an assertion). Any ideas or, should I move it back to `SILGenExpr.cpp`?

xedin replied:
I think `coeceCallArguments` does the right thing for keypath component, just like for other expressions with default arguments. Maybe it makes sense to check how regular subscript calls with default arguments are lowered to see if it does make sense to do the transformation in SILGen?

zoecarver replied:
Regular subscripts call `buildSubscriptHelper` which calls `coerceCallArguments` and that handles the default args. Interestingly, we also call `coeceCallArguments` but the default parameters aren't lowered. I'll look into why this is.

zoecarver replied:
I've moved the default argument lowering into [`coerceCallArguments`](https://github.com/apple/swift/pull/27951/files#diff-cf676abf2776cbcdb21380661ec41563R5403-R5411). It seems to work but, I am curious as to why this wasn't done before and suspect it may be because there are some cases where lowering the default args in `coerceCallArguments` doesn't work (that is why I kept the creation of `DefaultArgumentExpr` as a fallback).

zoecarver committed:
Refactor as suggested by @xedin

zoecarver committed:
Stash progress before changing branches

zoecarver committed:
Move default argument lowering into coerceCallArguments

zoecarver committed:
Clean up spacing and remove lingering debugging calls

xedin commented on lib/Sema/CSApply.cpp:
```diff
@@ -5400,10 +5400,15 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
-                                                 arg->getStartLoc(),
-                                                 param.getParameterType());
+        const ParamDecl *defaultParam = getParameterAt(cast<ValueDecl>(callee.getDecl()), newArgs.size());
+        defArg = defaultParam->getDefaultValue();
```
This would still be broken if default is something like `= #line` or other magic literal or `nil`. I'll take a look at how happens with default arguments in SILGen for the normal calls, maybe that'll give us some insights how to be about this...

zoecarver replied:
@xedin it doesn't seem to be an issue. I tested with both `subscript(x: Int = #line)` and `subscript(x: Int? = nil)`. I'll update the tests.

zoecarver committed:
Merge branch 'master' into key-path-default-arg

zoecarver committed:
Add edgecases to tests

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -5399,10 +5399,15 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
-            new (tc.Context) DefaultArgumentExpr(callee, paramIdx,
-                                                 arg->getStartLoc(),
-                                                 param.getParameterType());
+        const ParamDecl *defaultParam = getParameterAt(cast<ValueDecl>(callee.getDecl()), newArgs.size());
```
Isn't callee.getDecl() already a ValueDecl?

Also a nitpick: we try to keep lines < 80 columns

zoecarver replied:
Yep, fixed.

slavapestov commented on test/SILGen/keypaths.swift:
```diff
@@ -352,6 +385,19 @@ func subscripts<T: Hashable, U: Hashable>(x: T, y: U, s: String) {
+  _ = \SubscriptDefaults1.[bool: false, 0, 0]
+  
+  _ = \SubscriptDefaults2.[]
+  _ = \SubscriptDefaults3.[]
```
Just for completeness can you add some examples where the arguments are fully specified as well?

zoecarver replied:
Good idea.

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -5399,10 +5399,15 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
+        const ParamDecl *defaultParam = getParameterAt(cast<ValueDecl>(callee.getDecl()), newArgs.size());
+        defArg = defaultParam->getDefaultValue();
+
+        if (defArg == nullptr) {
```
It looks like defArg has to be nullptr at this point, because we're in the else branch of an `if (defArg)` check.

zoecarver replied:
`defArg` is set two lines above (line 5404).

slavapestov replied:
D'oh! My apologies.

zoecarver replied:
No worries :) Could you trigger the bot?

theblixguy replied:
There's a merge conflict on your PR

zoecarver committed:
Fix based on review comments

hamishknight commented on lib/Sema/CSApply.cpp:
```diff
@@ -5399,10 +5399,16 @@ Expr *ExprRewriter::coerceCallArguments(Expr *arg, AnyFunctionType *funcType,
-                                                 param.getParameterType());
+        const ParamDecl *defaultParam = getParameterAt(callee.getDecl(),
+                                                       newArgs.size());
+        defArg = defaultParam->getDefaultValue();
```
I'm not sure we want to be inlining regular default arguments into the resulting AST, as it may cause problems for things like autoclosures, which get contextualized to the parameter's context rather than the call-site's context. In addition, I don't believe this will work in the general case for default arguments in other modules, as we currently don't deserialize an expression for them.

For example:

```swift
// Module A
public struct S {
  public subscript(_ x: Int = 0) -> Int { x }
}

// Module B
import A

func bar() {
  _ = \S.[]
}
```

When coercing the call arguments for the key-path subscript, `defaultParam->getDefaultValue()` will be `nullptr`.

The way we usually handle default arguments is to emit a "default argument generator" per default argument in the same file they're declared in. Other modules that want to use the default arguments emit a call to this generator function, and use the result as the default value.

I think we'll want to take a similar approach for key-path subscript defaults, so we'll want to carry on using `DefaultArgumentExpr` to represent the argument in the type-checked AST, but modify SILGen to emit a call to the appropriate default argument generator when lowering the arguments for the subscript.

I'm not too familiar with SILGen, but I think you'll want to see if you can use `ArgEmitter` to emit the arguments for you, this should handle `DefaultArgumentExpr` as a "delayed argument", which you should be able to emit afterwards through `emitDelayedArguments`. Does this sound reasonable @slavapestov?

zoecarver replied:
Do you think it would be beneficial to add a test with two modules? Also, just read the part about `ArgEmitter`. I'll see if I can use that instead.

hamishknight replied:
@zoecarver If we don't already have a test for cross-module default args, it would definitely be worth adding one

hamishknight reviewed (commented):
Thanks for working on this! Apologies for not reviewing sooner.

zoecarver committed:
Emit default arguments in silgen

zoecarver committed:
Merge branch 'master' into key-path-default-arg

zoecarver committed:
Move default arg generation into SILGen

zoecarver committed:
Run code through clang-format

zoecarver commented:
@hamishknight Thanks for the review. I moved everything into `SILGenExpr`. I'll address your other comment shortly. 

I tried a few different methods for generating the default argument but, none of them worked out (for various reasons). I ended up adding a method called `generateDefaultArgsForKeyPathSubscript` which re-generates all the arguments. 

Side note: If the index expression is a tuple type with size zero and a default argument is added, that tuple will turn into a paren expression. The same goes if the index expression is a paren type and a new argument is added (it will turn into a tuple). I had some complex logic to conditionally do different things based on the old and new types but I ended up scratching that in favor of the current approach (which is much simpler).

zoecarver commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3532,6 +3537,70 @@ SILGenModule::emitKeyPathComponentForDecl(SILLocation loc,
+
+    (void)++index;
+  }
+
```
Thoughts on exiting early if we haven't found any default args?

zoecarver committed:
Remove lingering newline

zoecarver committed:
Re-enable tests (oops)

theblixguy commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3532,6 +3537,70 @@ SILGenModule::emitKeyPathComponentForDecl(SILLocation loc,
+
+    auto newIndexExpr =
+        TupleExpr::createImplicit(SGF.getASTContext(), newArgs, elementNames);
+    component.setIndexExpr(newIndexExpr);
```
This looks suspect to me - SILGen should be able to lower an AST without modifying it. If not, then it needs to be taught how to do it (or the AST you’re passing is not correct and should be changed in CSApply). However, I don’t think creating or modifying ASTs in the backend is the right thing to do.

zoecarver replied:
Yes, I think you're right. I'll see if I can find a way to do this differently.

zoecarver replied:
Aha! It's so simple. I've been looking at this code all day but I just saw exactly what needs to be changed. Update coming soon.

zoecarver committed:
Move default arg generation into lambda instead of modifing AST

zoecarver commented:
I moved everything into `lowerSubscriptOperands` and renamed the old `lowerSubscriptOperands` to `lowerOperands` (because it lowers all operands, not just subscript operands). This is a lot cleaner than my other solution and functionally almost identical (also has the added benefit of not modifying the AST in SILGen).

zoecarver committed:
Regenerate labels in CSApply instead of keeping two sizes

hamishknight commented on lib/Sema/CSApply.cpp:
```diff
@@ -4615,6 +4617,10 @@ namespace {
 
+      auto component =
+          KeyPathExpr::Component::forSubscriptWithPrebuiltIndexExpr(
+              ref, newIndexExpr, newLabels, resolvedTy, componentLoc, {});
```
`forSubscriptWithPrebuiltIndexExpr` doesn't appear to make a copy of the argument labels you pass it, so you'll need to make a copy with e.g `cs.getASTContext().AllocateCopy(newLabels)`. It might even be worth performing a copy in the `KeyPathExpr::Component` constructor to defensively avoid future memory bugs here.

Also, if you want, you could pass `conformances` into `forSubscriptWithPrebuiltIndexExpr` here, as long as it also gets copied.

zoecarver replied:
Maybe `forSubscriptWithPrebuiltIndexExpr` should be updated to use placement new and return a pointer. As is, it doesn't return a pointer so I don't think we can/should copy `component`.

hamishknight replied:
We don't want to be copying `component`, sorry what I meant was we could perform the copying operation within:

https://github.com/apple/swift/blob/cd3ada5abbe78d4062b7d323e5fd15a2df3d3524/lib/AST/Expr.cpp#L2145-L2163

rather than at the call-site. But given we only need to perform the copy at this one call-site, it's probably not worth it for now.

hamishknight commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3553,27 +3553,60 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
+    }
+  };
+
+  auto lowerOperands = [this, &operands,
```
You should be able to get rid of `lowerOperands`, as only subscript components have index expressions.

hamishknight commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3553,27 +3553,60 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
+    if (auto paren = dyn_cast<ParenExpr>(component.getIndexExpr())) {
+      auto param = decl->getIndices()->get(0);
+      if (param->isDefaultArgument())
+        indexResult = visit(param->getDefaultValue(), SGFContext());
```
Unfortunately, I believe this will run into the same problem as before with cross-module default arguments, as we don't deserialize an expression for them. Instead of trying to directly emit SIL for the expression, I think we should be trying to emit an apply of the appropriate default argument generator (see `DelayedArgument::emitDefaultArgument`). The benefit of using `ArgEmitter` is that it should handle this case for us, did you have any luck with that?

zoecarver replied:
I tried `SGF.SGM.emitDefaultArgGenerators` which tried to re-emit the subscript function. Then I tried to use `emitGeneratorFunction` directly which got further before having issues. Neither ended up working. I'll keep playing with it, though. 

Also, `emitDefaultArgGenerators` uses `param->getDefaultValue` which means it would have the same issue, right? Or is it OK because it calls `emitOrDelayFunction`?

zoecarver replied:
Actually, looking at [the code](https://github.com/apple/swift/blob/master/lib/SILGen/SILGen.cpp#L1062-L1073) again it looks like the call to `getDefaultValue` _isn't_ in the lambda which means it would be called around the same time either way.

hamishknight replied:
@zoecarver `emitDefaultArgGenerators` works because it's called when emitting the function declaration itself, and therefore has access to the default argument (e.g for a function with a default arg in Module A that gets used from Module B, `emitDefaultArgGenerators` only gets called when emitting Module A).

You shouldn't need to call it (or `emitGeneratorFunction`) when emitting default arguments at a call-site, as the appropriate default argument generator should have already been emitted. What we want to emit is an apply of that generator function (which is what `DelayedArgument::emitDefaultArgument` does).

zoecarver replied:
That makes sense, thanks for explaining. `DelayedArgument` only exists in `SILGenApply.cpp` and (AFAICT) it doesn't touch key paths. How should I access `DelayedArgument::emitDefaultArgument`?

hamishknight replied:
@zoecarver You'll probably want to break off a function that deals with key path subscript argument lowering into SILGenApply.cpp so it has access to those functions. If you're able to use `ArgEmitter`, then it'll give you a list of delayed arguments which you can emit by calling `emitDelayedArguments`, which will call through to `DelayedArgument::emitDefaultArgument` for you.

zoecarver replied:
I added a function in SILGenApply.cpp where I call `emitSingleArg` via `ArgEmitter` for every element in an array of collected index expressions. But, it still errors when it receives a `DefaultArgumentExpr`. Any ideas (I can push my code if that would be helpful)?

Also, I noticed that `visitSubscriptDecl` calls `emitDefaultArgGenerators`. Maybe that should be updated to use `ArgEmitter`  as well. I'll see if I can test it with multiple modules and see if there's an issue. 

Thanks for all the help with this :)

hamishknight replied:
@zoecarver I think you may want to use `PreparedArguments` in order to decompose the tuple/paren expr for you, which you can then pass to `ArgEmitter::emitPreparedArgs`. Actually I've just noticed we have `SILGenFunction::prepareSubscriptIndices` which is used to decompose arguments for regular subscripts, I wonder if we can re-use that here?

Regarding `visitSubscriptDecl`, note that this is for the subscript declaration, not the call-site. `SILGenLValue::visitSubscriptExpr` is where you want to look for regular subscript emission. The argument emission logic is a bit convoluted, but it'll eventually wind up calling various routines depending on the exact accessor being used, e.g `SILGenFunction::emitGetAccessor`. This logic uses `CallEmission` which will use `ArgEmitter` for the arguments (I don't believe we can use `CallEmission` ourselves as we don't want an actual call, we just want the arguments).

zoecarver replied:
Updated to use `ArgEmitter`! This definitely feels like the right way to do it (and it makes all argument parsing more general).

hamishknight commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -3553,27 +3553,60 @@ RValue RValueEmitter::visitKeyPathExpr(KeyPathExpr *E, SGFContext C) {
+      return;
+
+    auto decl = dyn_cast<SubscriptDecl>(component.getDeclRef().getDecl());
+    assert(decl &&
```
Nit: You can use `cast` instead of `dyn_cast` + `assert`

hamishknight reviewed (commented):
Definitely heading in the right direction!

zoecarver committed:
Use ArgEmitter

zoecarver committed:
Create emitKeyPathSubscriptOperands to lower all the args properly (including the default ones)

zoecarver committed:
Use range-based for-loop

zoecarver commented:
After running the tests, it seems there is an issue when `ArgEmitter` tries to emit args with generic types. For example:
```swift
struct Foo<T : Numeric> {
  subscript(x: T = 0) -> T { x }
}

_ = \Foo<Int32>.[]
```
I'll try to work this out tomorrow, but if anyone knows what might be causing the problem, please let me know.

hamishknight commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -6067,6 +6067,39 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
 
+SmallVector<ManagedValue, 4> SILGenFunction::emitKeyPathSubscriptOperands(
+    SubscriptDecl *subscript, SubstitutionMap subs, Expr *indexExpr,
+    CanSILFunctionType fnType) {
```
I believe the problem you're running into with generic parameters is due to the fact that you're passing the function type for the key path accessor here, which will contain generic parameter types. However the the original parameter types you're supplying (`origFnType`) have been substituted with the replacement types.

As far as I can tell, we just want to be emitting the arguments directly without performing any abstraction changes or things like that, so I *think* what we want to do is remove the `fnType` parameter here, and lower the concrete type of the function ourselves, e.g `auto fnType = getLoweredType(origFnType, substFnType).castTo<SILFunctionType>()`. That appears to work at least for the generic example you gave.

zoecarver replied:
Fantastic! This seems to pass all the tests.

hamishknight replied:
Great! Could you add some FileCheck SILGen tests alongside your existing tests in test/SILGen/keypaths.swift to make sure we emit the correct SILGen for a few cases? I think we'll want a test for at least:

- Your example with a generic default
- An example with a caller-side default such as `#function`
- An example with a generic caller-side default, e.g `subscript<T : ExpressibleByStringLiteral>(x x: T = #function, y y: T) -> T { x }`, called using `\Foo.[y: ""]`

You can take a look at:

https://github.com/apple/swift/blob/cd3ada5abbe78d4062b7d323e5fd15a2df3d3524/test/SILGen/default_arguments.swift#L44-L57

to see an example of a FileCheck test for the SILGen for a regular default argument.

It would be great if you could also add a test case for a cross-module default.

zoecarver replied:
I tried to write a FileCheck test but it can't match the string. I suspect its a simple issue. Any ideas?
**Input**
```swift
// CHECK-LABEL: sil hidden [ossa] @{{.*}}check_default_subscripts
func check_default_subscripts() {
  // CHECK: [[INTINIT:%[0-9]+]] = integer_literal $Builtin.Int64, 0
  // CHECK: [[I:%[0-9]+]] = struct $Int ([[INTINIT]] : $Builtin.Int64)
  // CHECK: [[DFN:%[0-9]+]] = function_ref @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluipfA0_ : $@convention(thin) <τ_0_0 where τ_0_0 : Numeric> () -> @out τ_0_0
  // CHECK: [[ALLOC:%[0-9]+]] = alloc_stack $Int
  // CHECK: apply [[DFN]]<Int>([[ALLOC]]) : $@convention(thin) <τ_0_0 where τ_0_0 : Numeric> () -> @out τ_0_0
  // CHECK: [[LOAD:%[0-9]+]] = load [[ALLOC]] : $*Int
  // CHECK: [[KEYPATH:%[0-9]+]] = keypath $WritableKeyPath<SubscriptDefaults4, Int>, (root $SubscriptDefaults4; settable_property $Int,  id @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluig : $@convention(method) <τ_0_0 where τ_0_0 : Numeric> (@in_guaranteed τ_0_0, @in_guaranteed τ_0_0, SubscriptDefaults4) -> @out τ_0_0, getter @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluipACSiTK : $@convention(thin) (@in_guaranteed SubscriptDefaults4, UnsafeRawPointer) -> @out Int, setter @$s3run18SubscriptDefaults4V1x1yxx_xtcSjRzluipACSiTk : $@convention(thin) (@in_guaranteed Int, @inout SubscriptDefaults4, UnsafeRawPointer) -> (), indices [%$0 : $Int : $Int, %$1 : $Int : $Int], indices_equals @$sS2iTH : $@convention(thin) (UnsafeRawPointer, UnsafeRawPointer) -> Bool, indices_hash @$sS2iTh : $@convention(thin) (UnsafeRawPointer) -> Int) ([[I]], [[LOAD]])
  _ = \SubscriptDefaults4.[x: 0]
}
```
**Error**
```
swift/test/SILGen/keypaths.swift:423:12: error: CHECK: expected string not found in input
 // CHECK: [[INTINIT:%[0-9]+]] = integer_literal $Builtin.Int64, 0
           ^
<stdin>:5700:57: note: scanning from here
sil hidden [ossa] @$s8keypaths24check_default_subscriptsyyF : $@convention(thin) () -> () {
                                                        ^
<stdin>:5702:2: note: possible intended match here
 %0 = integer_literal $Builtin.IntLiteral, 0 // user: %3
```
I've looked and the emitted silgen _does_ contian `integer_literal $Builtin.Int64, 0`:
```
// check_default_subscripts()
sil hidden @$s3run24check_default_subscriptsyyF : $@convention(thin) () -> () {
bb0:
  %0 = integer_literal $Builtin.Int64, 0          // user: %1
  %1 = struct $Int (%0 : $Builtin.Int64)          // user: %6
...
```

Thanks again for the help!

hamishknight replied:
@zoecarver Ah, I think you're looking at output from `-emit-sil` instead of `-emit-silgen`, the test uses the latter so that's what you'll want to use for reference.

zoecarver replied:
@hamishknight yep, that's it!

zoecarver replied:
@hamishknight sorry for the delay. I think I've got all the test cases you asked for. Let me know if I missed any, though.

zoecarver commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -6067,6 +6067,39 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
+  SmallVector<DelayedArgument, 2> delayedArgs;
+  ArgEmitter emitter(
+      *this, fnType.getPointer()->getRepresentation(),
+      /*yield*/ true,
```
Yep, copy-paste error.

zoecarver committed:
Use correct function type to prevent conflicting argument types

zoecarver committed:
Format properly

zoecarver committed:
Copy labels and conformances

zoecarver committed:
Fix spacing

zoecarver committed:
Fix spacing

zoecarver committed:
Fix spacing (again)

zoecarver committed:
Fix merge conflict

hamishknight commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -6067,6 +6067,40 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
+  SmallVector<ManagedValue, 4> argValues;
+  SmallVector<DelayedArgument, 2> delayedArgs;
+  ArgEmitter emitter(
+      *this, fnType.getPointer()->getRepresentation(),
```
Nit: You can do `fnType->getRepresentation()` here

hamishknight commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -6067,6 +6067,40 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
+      *this, fnType.getPointer()->getRepresentation(),
+      /*yield*/ false,
+      /*isForCoroutine*/ false,
+      ClaimedParamsRef(fnType, fnType.getPointer()->getParameters()), argValues,
```
Same here

hamishknight commented on lib/SILGen/SILGenFunction.h:
```diff
@@ -664,7 +664,11 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
-  
+
+  SmallVector<ManagedValue, 4>
+  emitKeyPathSubscriptOperands(SubscriptDecl *subscript, SubstitutionMap subs,
```
Could you add a brief doc comment here to explain that this function emits arguments for a key path subscript component?

hamishknight reviewed (approved):
Thanks for sticking with this! As already mentioned, I think we should add a few FileCheck SILGen tests to make sure we're emitting things correctly, but otherwise this looks good to me.

I would like someone with more SILGen experience to have a look though, @jckarter would you mind taking a look?

zoecarver committed:
Add FileCheck tests

zoecarver committed:
Add tests

zoecarver committed:
More tests

zoecarver committed:
Format with clang-format

zoecarver commented:
Could someone trigger the bot, please?

hamishknight commented on lib/SILGen/SILGenFunction.h:
```diff
@@ -664,7 +664,19 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
+  /// \returns Lowered index arguments.
+  /// \param subscript - The subscript decl who's arguments are being lowered.
+  /// \param subs - Used to get subscript function type and to substitute
+  /// generic args. \param indexExpr - An expression holding the indices of the
```
Nit: Could you put `\param indexExpr` on a new line?

hamishknight commented on lib/SILGen/SILGenFunction.h:
```diff
@@ -664,7 +664,19 @@ class LLVM_LIBRARY_VISIBILITY SILGenFunction
+  /// \param subscript - The subscript decl who's arguments are being lowered.
+  /// \param subs - Used to get subscript function type and to substitute
+  /// generic args. \param indexExpr - An expression holding the indices of the
+  /// subscript (either a TupleExpr or a ParanExpr).
```
Typo: ParanExpr -> ParenExpr

hamishknight commented on test/SILGen/keypaths.swift:
```diff
@@ -352,6 +399,45 @@ func subscripts<T: Hashable, U: Hashable>(x: T, y: U, s: String) {
+  _ = \SubscriptDefaults3.[0]
+  _ = \SubscriptDefaults4.[x: 0]
+  _ = \SubscriptDefaults4.[x: 0, y: 0]
+  _ = \SubscriptDefaults5.[x: ""]
```
Could you make `\SubscriptDefaults5.[x: ""]` a FileCheck test? Just to make sure we correctly handle the interaction between generic params and caller-side defaults.

hamishknight reviewed (commented):
Thanks for adding those tests! If you could add one extra FileCheck test for `\SubscriptDefaults5.[x: ""]`, I'll go ahead and kick off the bot.

zoecarver committed:
Add more FileCheck tests

zoecarver committed:
Fix nits

zoecarver commented:
@hamishknight all fixed!

hamishknight commented:
@zoecarver Thanks!

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

hamishknight commented:
Just to throw some extra code at it

@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test OS X platform

slavapestov commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -6067,6 +6067,39 @@ RValue SILGenFunction::emitDynamicSubscriptExpr(DynamicSubscriptExpr *e,
+
+  SmallVector<ManagedValue, 4> argValues;
+  SmallVector<DelayedArgument, 2> delayedArgs;
+  ArgEmitter emitter(*this, fnType->getRepresentation(),
```
Nice! Using ArgEmitter is definitely the right approach here.

It appears that the old code didn't correctly handle keypaths with variadic subscript components either, eg

```
struct S {
  subscript<T>(x: T...) -> Int { get {} set {} }
}

let k = \S.[1, 2, 3]
```

Crashes with this on master

```
Assertion failed: (hashables.size() == SubscriptSize), function setSubscriptIndexHashableConformances, file /Users/spestov/src/swift/lib/AST/Expr.cpp, line 2129.
Stack dump:
0.	Program arguments: /Users/spestov/src/build/Ninja-RelWithDebInfoAssert/swift-macosx-x86_64/bin/swift -frontend -c -primary-file var.swift -target x86_64-apple-darwin19.0.0 -enable-objc-interop -color-diagnostics -module-name var -o /var/folders/wg/nmvf884n4b14lt0yw4khjsv00000gn/T/var-9e61ce.o 
1.	Swift version 5.1.1-dev (Swift 69c2de1727)
2.	While type-checking statement at [var.swift:5:1 - line:5:20] RangeText="let k = \S.[1, 2, 3"
3.	While type-checking declaration 0x7fd9ed0ca3a0 (at var.swift:5:1)
4.	While evaluating request PatternBindingEntryRequest((unknown decl), 0)
5.	While type-checking expression at [var.swift:5:9 - line:5:20] RangeText="\S.[1, 2, 3"
```

It looks like a similar issue. Can you see if your fix addresses this as well? If not, a follow up PR would be great :)

zoecarver replied:
I'll have to create a follow-up patch for variadics as it looks like that will require some additional work.

slavapestov commented on lib/Sema/CSApply.cpp:
```diff
@@ -4615,8 +4615,10 @@ namespace {
-      component.setSubscriptIndexHashableConformances(conformances);
-      return component;
+      return KeyPathExpr::Component::forSubscriptWithPrebuiltIndexExpr(
+          ref, newIndexExpr, cs.getASTContext().AllocateCopy(newLabels),
```
It looks like KeyPathExpr::Component::getSubscriptLabels() is only called for KeyPathExpr::Component::Kind::UnresolvedSubscript. It might be worth a try to stick an assertion there if the kind is anything else, and remove the parameter from forSubscriptWithPrebuiltIndexExpr() altogether.

slavapestov commented:
> My only lingering question is if anyone has objections to removing the call to relabelParams. It seems like it does nothing.

I *just* noticed this in your PR description. I believe relabelParams() is a holdover from before some refactoring I did that changed the representation of calls in SILGen. You can remove it.

zoecarver commented:
@slavapestov it's from an earlier version of this patch. Now I'm not touching that part of the code. I'll look into creating another patch to resolve that and clean up `forSubscriptWithPrebuiltIndexExpr`. Thanks for merging this!
